Feature/category

PadillaTom opened the pull request:
### Summary

Tareas:
- OT111-42 -> POST newCategory
- OT111-41 -> GET Details by Id
- OT111-40 -> GET All Categories Name
- OT111-44 -> DELETE by Id
- OT111-43 ->PUT Category by Id

![image](https://user-images.githubusercontent.com/59969816/145558401-dd0db4b8-3a32-4527-9fdb-1c23fd22efbc.png)

### Evidence
POST
![image](https://user-images.githubusercontent.com/59969816/145559032-34483327-46a8-4030-939c-ccba9826341a.png)

GET Details
![image](https://user-images.githubusercontent.com/59969816/145559122-fd00b835-0203-40e5-aa45-c2599f56fbbb.png)

GET All Categories
![image](https://user-images.githubusercontent.com/59969816/145559332-2bc08ca3-56ab-4a4c-ba5f-2b0476efc8a0.png)

DELETE by Id
![image](https://user-images.githubusercontent.com/59969816/145559492-fa6da47f-49d7-443e-9995-510506548457.png)

PUT by Id
![image](https://user-images.githubusercontent.com/59969816/145559739-01b94dc1-707b-43b7-a6ce-e7e1d041c85e.png)
![image](https://user-images.githubusercontent.com/59969816/145559701-64db9a2a-c015-42dd-95c5-2300667207b3.png)

PadillaTom committed:
Category - Created Branch

PadillaTom committed:
Feature/ot111 42 (#29)

* OT111-42 - POST Category OK con response

* OT111-42 - POST solo por Admin OK

* OT111-42 - Added ID to CategoryResponseDTO

PadillaTom committed:
Feature/ot111 41 (#31)

* OT111-41 - GET solo por ADMIN OK, falta return 404

* OT111-41 - Added ID to CategoryResponseDTO, falta 404

* OT111-41 - GET by Admin o 404 OK

PadillaTom committed:
Merge branch 'develop' into feature/category

PadillaTom committed:
Category - Merged Develop, Fixed with Sonar

PadillaTom committed:
Merge branch 'develop' into feature/category

danit3x committed:
OT111-40v2 Endpoint Listado Categories (#37)

PadillaTom committed:
Merge branch 'develop' into feature/category

PadillaTom committed:
Merge branch 'feature/category' of https://github.com/alkemyTech/OT111-Server into feature/category

PadillaTom committed:
Category - Get List, Create, GetByID OK

wallyjj277 committed:
Feature/ot111 44 endpoint eliminar-categoriasv2 (#45)

* modifico controlador para el metodo borrar, modifico el service para el metodo borrar, modifico el categori service impl para el metodo borrar, hay errores a revisar en el metodo borrar

* Ot111-44/endpoint delete category

modifico controller endpoint, modifico el metodo delete, soluciono errores, agrego una linea para permitir acceso en security solo admin para delete

* Update CategoryController.java

cambio el return que daba error

PadillaTom committed:
Category - Approved DELETE by ID

wallyjj277 committed:
Feature/ot111 43 endpoint- actualizar-Categories (#46)

* modifico controlador para el metodo borrar, modifico el service para el metodo borrar, modifico el categori service impl para el metodo borrar, hay errores a revisar en el metodo borrar

* Ot111-44/endpoint delete category

modifico controller endpoint, modifico el metodo delete, soluciono errores, agrego una linea para permitir acceso en security solo admin para delete

* Update CategoryController.java

cambio el return que daba error

* OT111-43 endpoint-actualizar

se añade rol admin para el endpoint actualizar, se crea el metodo put en el controller, se crea el metodo en el servicio para actualizar, probado funcionando con postman

Co-authored-by: Tom <59969816+PadillaTom@users.noreply.github.com>

denisdimarco commented on src/main/java/com/alkemy/ong/service/impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,79 @@
+    @Override
+    @Transactional
+    public boolean deleteCategory(Long id){
+        try {
+            Optional<CategoryEntity>categoryEntityOptional = categoryRepository.findById(id);
+            CategoryEntity category = categoryEntityOptional.get();
+            category.setDeleted(true);
+            categoryRepository.save(category);
+            return true;
```
Estamos cambiando manualmente el deleted a true para borrar, pero ya estaba activado el soft delete en **CategoryEntity** con:
```java
@SQLDelete(sql = "UPDATE category SET deleted = true WHERE id=?")
@Where(clause = "deleted = false")
```
 podriamos entonces usar el metedo directamente deleteById(id) para hacer soft delete:
```java
categoryRepository.deleteById(id);
```

wallyjj277 replied:
se cambio el metodo por uno funcional. resuelto

danit3x committed:
Fix delete method

OsvaldoGWP commented on src/main/java/com/alkemy/ong/controller/CategoryController.java:
```diff
@@ -0,0 +1,59 @@
+        if (categoryDetails == null) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
+        }
+        return ResponseEntity.status(HttpStatus.ACCEPTED).body(categoryDetails);
```
deberia retornar 200 cuando es exitoso, chequear http status convencion

wallyjj277 replied:
se cambio el ACCEPTED POR OK AHORA RETORNA 200

OsvaldoGWP commented on src/main/java/com/alkemy/ong/controller/CategoryController.java:
```diff
@@ -0,0 +1,59 @@
+        if (!isDeleted) {
+            return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
+        }
+        return new ResponseEntity<>(HttpStatus.OK);
```
para el de4leted vamos a usar codigo 204

wallyjj277 replied:
SE CAMBIO EL  CODIGO 200 POR UN 204

OsvaldoGWP commented on src/main/java/com/alkemy/ong/service/CategoryService.java:
```diff
@@ -0,0 +1,21 @@
+
+public interface CategoryService {
+
+    CategoryResponseDTO findCategory(Long id);
```
quizas findCategoryBy o findCategoryById

wallyjj277 replied:
SE CAMBIO EL NOMBRE DEL METODO POR UNO MAS ESPECIFICO FIND BY ID

OsvaldoGWP commented on src/main/java/com/alkemy/ong/service/impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+    @Override
+    @Transactional
+    public boolean deleteCategory(Long id){
+        try {
```
chequear a futuro, agregar un todo

wallyjj277 replied:
SE AGREGA UN TODO PARA CHEQUEAR A FUTURO POSIBLES MEJORAS

OsvaldoGWP commented on src/main/java/com/alkemy/ong/service/impl/CategoryServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+    @Override
+    public CategoryResponseDTO updateCategory(CategoryRequestDTO request, Long id) throws Exception {
+        Optional<CategoryEntity>categoryEntityOptional = categoryRepository.findById(id);
+        if (!categoryEntityOptional.isPresent()){
```
usar .isEmpty asi no negarlo

wallyjj277 replied:
SE CAMBIO LOS IS PRESENT POR IS EMPTY EN LOS CASOS QUE REQUERIA

wallyjj277 committed:
fix de categories

se cambian el metodo delete, se cambia la devolucion de ok, no content, y arreglos menores.

wallyjj277 committed:
Update CategoryServiceImpl.java

SE AGREGA EL TODO REQUERIDO POR LA REVIEW DEL PR

wallyjj277 committed:
Merge branch 'develop' into feature/category
